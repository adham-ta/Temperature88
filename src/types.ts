import express from "express";
import { WebhookEvent, Webhooks } from "@octokit/webhooks";
import LRUCache from "lru-cache";
import Redis from "ioredis";

import { Probot } from "./index";
import { Context } from "./context";
import { ProbotOctokit } from "./octokit/probot-octokit";
import { Application } from "./application";

import type { Logger, LogFn } from "pino";

export interface Options {
  // same options as Application class
  privateKey?: string;
  githubToken?: string;
  appId?: number | string;

  Octokit?: typeof ProbotOctokit;
  log?: Logger;
  redisConfig?: Redis.RedisOptions | string;
  secret?: string;
  webhookPath?: string;
  logLevel?: "trace" | "debug" | "info" | "warn" | "error" | "fatal";
  port?: number;
  host?: string;
  webhookProxy?: string;
  baseUrl?: string;

  // Probot class-specific options
  /**
   * @deprecated `id` options is deprecated. Use `appId` instead
   */
  id?: number | string;
  /**
   * @deprecated `cert` options is deprecated. Use `privateKey` instead
   */
  cert?: string;
  /**
   * @deprecated set `Octokit` to `ProbotOctokit.defaults({ throttle })` instead
   */
  throttleOptions?: any;
}

export type State = {
  appId?: number;
  privateKey?: string;
  githubToken?: string;
  log: Logger;
  Octokit: typeof ProbotOctokit;
  octokit: InstanceType<typeof ProbotOctokit>;
  cache?: LRUCache<number, string>;
  webhooks: {
    path?: string;
    secret?: string;
  };
  port?: number;
  host?: string;
  webhookProxy?: string;
  baseUrl?: string;
};

export type ProbotWebhooks = Webhooks<
  WebhookEvent,
  Omit<Context, keyof WebhookEvent>
>;

export type DeprecatedLogger = LogFn & Logger;

type deprecatedKeys =
  | "router"
  | "log"
  | "on"
  | "receive"
  | "load"
  | "route"
  | "auth";

export type ApplicationFunctionOptions = {
  /**
   * @deprecated "(app) => {}" is deprecated. Use "({ app }) => {}" instead.
   */
  [K in deprecatedKeys]: Application[K];
} & { app: Probot; getRouter: (path?: string) => express.Router };
export type ApplicationFunction = (options: ApplicationFunctionOptions) => void;

export type ServerOptions = {
  log?: Logger;
  port?: number;
  host?: string;
  webhookPath?: string;
  webhookProxy?: string;
  Probot: typeof Probot;
};

export type MiddlewareOptions = {
  probot: Probot;
  [key: string]: unknown;

  /**
   * @deprecated "Probot" option is deprecated. Pass a "probot" instance instead, see https://github.com/probot/probot/pull/1431
   */
  Probot?: typeof Probot;
};
