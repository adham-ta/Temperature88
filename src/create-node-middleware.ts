import { RequestListener } from "http";

import { Deprecation } from "deprecation";

import { ApplicationFunction } from "./types";
import { MiddlewareOptions } from "./types";

export function createNodeMiddleware(
  appFn: ApplicationFunction,
  { probot, Probot }: MiddlewareOptions
): RequestListener {
  if (Probot) {
    probot = new Probot();
    probot.log.warn(
      new Deprecation(
        `"createNodeMiddleware(app, { Probot })" is deprecated. Use "createNodeMiddleware(app, { probot })" instead`
      )
    );
  }

  probot.load(appFn);

  return probot.webhooks.middleware;
}
