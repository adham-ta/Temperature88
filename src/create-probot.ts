import { Deprecation } from "deprecation";
import { LogLevel, Options as PinoOptions } from "@probot/pino";
import { getPrivateKey } from "@probot/get-private-key";

import { getLog, GetLogOptions } from "./helpers/get-log";
import { Options } from "./types";
import { Probot } from "./probot";

type CreateProbotOptions = {
  overrides?: Options;
  defaults?: Options;
  env?: NodeJS.ProcessEnv;
};

const DEFAULTS = {
  APP_ID: "",
  WEBHOOK_SECRET: "",
  GHE_HOST: "",
  GHE_PROTOCOL: "",
  LOG_FORMAT: "",
  LOG_LEVEL: "warn",
  LOG_LEVEL_IN_STRING: "",
  REDIS_URL: "",
  SENTRY_DSN: "",
};

/**
 * Merges configuration from defaults/environment variables/overrides and returns
 * a Probot instance. Finds private key using [`@probot/get-private-key`](https://github.com/probot/get-private-key).
 *
 * @see https://probot.github.io/docs/configuration/
 * @param defaults default Options, will be overwritten if according environment variable is set
 * @param overrides overwrites defaults and according environment variables
 * @param env defaults to process.env
 */
export function createProbot(options: Options | CreateProbotOptions = {}) {
  if (isDeprecated(options)) {
    return deprecatedCreateProbot(options);
  }

  const { overrides = {}, defaults = {}, env = process.env } = options;

  const privateKey = getPrivateKey({ env });
  const envWithDefaults = { ...DEFAULTS, ...env };

  const envOptions: Options = {
    logLevel: envWithDefaults.LOG_LEVEL as LogLevel,
    appId: Number(envWithDefaults.APP_ID),
    privateKey: (privateKey && privateKey.toString()) || undefined,
    secret: envWithDefaults.WEBHOOK_SECRET,
    redisConfig: envWithDefaults.REDIS_URL,
    baseUrl: envWithDefaults.GHE_HOST
      ? `${envWithDefaults.GHE_PROTOCOL || "https"}://${
          envWithDefaults.GHE_HOST
        }/api/v3`
      : "https://api.github.com",
  };

  const probotOptions = {
    ...defaults,
    ...envOptions,
    ...overrides,
  };

  const logOptions: GetLogOptions = {
    level: probotOptions.logLevel,
    logFormat: envWithDefaults.LOG_FORMAT as PinoOptions["logFormat"],
    logLevelInString: envWithDefaults.LOG_LEVEL_IN_STRING === "true",
    sentryDsn: envWithDefaults.SENTRY_DSN,
  };

  const log = getLog(logOptions).child({ name: "server" });

  return new Probot({
    log: log.child({ name: "probot" }),
    ...probotOptions,
  });
}

function isDeprecated(
  options: Options | CreateProbotOptions
): options is Options {
  const keys = Object.keys(options);

  return (
    keys.length > 0 &&
    !keys.includes("overrides") &&
    !keys.includes("defaults") &&
    !keys.includes("env")
  );
}

function deprecatedCreateProbot(options: Options) {
  options.log =
    options.log ||
    getLog({
      level: process.env.LOG_LEVEL as LogLevel,
      logFormat: process.env.LOG_FORMAT as PinoOptions["logFormat"],
      logLevelInString: process.env.LOG_LEVEL_IN_STRING === "true",
      sentryDsn: process.env.SENTRY_DSN,
    });

  const deprecatedKey = Object.keys(options).join(", ");
  options.log.warn(
    new Deprecation(
      `[probot] "createProbot({ ${deprecatedKey} })" is deprecated, use "new Probot(options)" instead.
      
"createProbot(options)" will be repurposed with probot v11 and only accept {defaults, overrides, env} option keys`
    )
  );
  return new Probot(options);
}
