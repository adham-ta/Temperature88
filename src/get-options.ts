import { Deprecation } from "deprecation";
import { getPrivateKey } from "@probot/get-private-key";
import { Options as PinoOptions, LogLevel } from "@probot/pino";

import { getLog, GetLogOptions } from "./helpers/get-log";
import { Probot } from "./probot";
import { Options, ServerOptions } from "./types";

type Parameters = {
  overrides?: Partial<ServerOptions>;
  defaults?: Partial<ServerOptions>;
  env?: NodeJS.ProcessEnv;
};

const DEFAULTS = {
  APP_ID: "",
  GHE_HOST: "",
  GHE_PROTOCOL: "",
  HOST: "",
  LOG_FORMAT: "",
  LOG_LEVEL: "warn",
  LOG_LEVEL_IN_STRING: "",
  PORT: "3000",
  REDIS_URL: "",
  SENTRY_DSN: "",
  WEBHOOK_PATH: "",
  WEBHOOK_PROXY_URL: "",
  WEBHOOK_SECRET: "",
};

/**
 * Merges configuration from defaults, environment variables, and overrides.
 * Finds private key using [`@probot/get-private-key`](https://github.com/probot/get-private-key).
 *
 * @see https://probot.github.io/docs/configuration/
 * @param defaults default Options, will be overwritten if according environment variable is set
 * @param overrides overwrites defaults and according environment variables
 * @param env defaults to process.env
 */
export function getOptions(
  {
    overrides = {},
    defaults = {},
    env = process.env,
  }: Parameters = {} as Parameters
): ServerOptions {
  const privateKey = getPrivateKey({ env });
  const envWithDefaults = { ...DEFAULTS, ...env };

  const logOptions: GetLogOptions = {
    level: envWithDefaults.LOG_LEVEL as LogLevel,
    logFormat: envWithDefaults.LOG_FORMAT as PinoOptions["logFormat"],
    logLevelInString: envWithDefaults.LOG_LEVEL_IN_STRING === "true",
    sentryDsn: envWithDefaults.SENTRY_DSN,
  };

  const probotOptions: Options = {
    appId: Number(envWithDefaults.APP_ID),
    privateKey: (privateKey && privateKey.toString()) || undefined,
    secret: envWithDefaults.WEBHOOK_SECRET,
    redisConfig: envWithDefaults.REDIS_URL,
    baseUrl: envWithDefaults.GHE_HOST
      ? `${envWithDefaults.GHE_PROTOCOL || "https"}://${
          envWithDefaults.GHE_HOST
        }/api/v3`
      : "https://api.github.com",
  };

  const log = getLog(logOptions).child({ name: "server" });

  const ProbotWithDefaults = Probot.defaults({
    ...probotOptions,
    log: log.child({ name: "probot" }),
  });

  const options = {
    ...defaults,
    host: envWithDefaults.HOST,
    port: Number(envWithDefaults.PORT),
    webhookPath: envWithDefaults.WEBHOOK_PATH,
    webhookProxy: envWithDefaults.WEBHOOK_PROXY_URL,
    log,
    Probot: ProbotWithDefaults,
    ...overrides,
  };

  options.log.warn(
    new Deprecation(
      `[probot] "getOptions()" is deprecated, use "{ probot: createProbot() }" instead:

    const { createNodeMiddleware, createProbot } = require("probot");
    const myApp = require("./my-app.js");

    module.exports = createNodeMiddleware(myApp, { probot: createProbot() });`
    )
  );

  return options;
}
