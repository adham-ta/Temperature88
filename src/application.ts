import { Router } from "express";
import Redis from "ioredis";
import LRUCache from "lru-cache";
import { Deprecation } from "deprecation";
import { Webhooks } from "@octokit/webhooks";
import { Logger } from "pino";
import { LogLevel } from "@probot/pino";

import { Context } from "./context";
import { ProbotOctokit } from "./octokit/probot-octokit";
import { getLog } from "./helpers/get-log";
import { getProbotOctokitWithDefaults } from "./octokit/get-probot-octokit-with-defaults";
import {
  DeprecatedLogger,
  ProbotWebhooks,
  State,
  ApplicationFunction,
} from "./types";
import { webhookEventCheck } from "./helpers/webhook-event-check";
import { aliasLog } from "./helpers/alias-log";
import { getWebhooks } from "./octokit/get-webhooks";
import { load } from "./load";
import { getRouter } from "./get-router";
import { auth } from "./auth";

export interface Options {
  // same options as Probot class
  privateKey?: string;
  githubToken?: string;
  id?: number;
  Octokit?: typeof ProbotOctokit;
  log?: Logger;
  redisConfig?: Redis.RedisOptions;
  secret?: string;
  webhookPath?: string;

  // Application class specific options
  cache?: LRUCache<number, string>;
  octokit?: InstanceType<typeof ProbotOctokit>;
  webhooks?: Webhooks;

  /**
   * @deprecated "app.router" is deprecated, use "getRouter()" from the app function instead: "({ app, getRouter }) => { ... }"
   */
  router?: Router;
  /**
   * @deprecated set `Octokit` to `ProbotOctokit.defaults({ throttle })` instead
   */
  throttleOptions?: any;
}

export type OnCallback<T> = (context: Context<T>) => Promise<void>;

/**
 * The `app` parameter available to `ApplicationFunction`s
 *
 * @property {logger} log - A logger
 */
export class Application {
  public log: DeprecatedLogger;
  public on: ProbotWebhooks["on"];
  public receive: ProbotWebhooks["receive"];
  public load: (
    appFn: ApplicationFunction | ApplicationFunction[]
  ) => Application;
  public auth: (
    installationId?: number,
    log?: Logger
  ) => Promise<InstanceType<typeof ProbotOctokit>>;

  private webhooks: ProbotWebhooks;
  private state: State;
  private internalRouter: Router;

  constructor(options: Options) {
    this.log = aliasLog(
      options.log || getLog({ level: process.env.LOG_LEVEL as LogLevel })
    );

    this.log.warn(
      `[probot] "import { Application } from 'probot'" is deprecated. Use "import { Probot } from 'probot'" instead, the APIs are the same.`
    );

    const cache =
      options.cache ||
      new LRUCache<number, string>({
        // cache max. 15000 tokens, that will use less than 10mb memory
        max: 15000,
        // Cache for 1 minute less than GitHub expiry
        maxAge: 1000 * 60 * 59,
      });

    const Octokit = getProbotOctokitWithDefaults({
      githubToken: options.githubToken,
      Octokit: options.Octokit || ProbotOctokit,
      appId: options.id,
      privateKey: options.privateKey,
      cache,
      log: this.log,
      redisConfig: options.redisConfig,
      throttleOptions: options.throttleOptions,
    });

    if (options.throttleOptions) {
      this.log.warn(
        `[probot] "new Application({ throttleOptions })" is deprecated. Use "new Application({Octokit: ProbotOctokit.defaults({ throttle }) })" instead`
      );
    }

    this.state = {
      cache,
      githubToken: options.githubToken,
      log: this.log,
      Octokit,
      octokit: options.octokit || new Octokit(),
      webhooks: {
        path: options.webhookPath,
        secret: options.secret,
      },
    };

    this.webhooks = options.webhooks || getWebhooks(this.state);

    this.on = (eventNameOrNames, callback) => {
      // when an app subscribes to an event using `app.on(event, callback)`, Probot sends a request to `GET /app` and
      // verifies if the app is subscribed to the event and logs a warning if it is not.
      //
      // This feature will be moved out of Probot core as it has side effects and does not work in a stateless environment.
      webhookEventCheck(this.state, eventNameOrNames);

      if (eventNameOrNames === "*") {
        // @ts-ignore this workaround is only to surpress a warning. The `.on()` method will be deprecated soon anyway.
        return this.webhooks.onAny(callback);
      }

      return this.webhooks.on(eventNameOrNames, callback);
    };
    this.receive = this.webhooks.receive;

    const router = options.router || Router();
    // @ts-ignore
    this.load = load.bind(null, this, router);
    this.auth = auth.bind(null, this.state);

    this.internalRouter = router;
  }

  /**
   * @deprecated "app.router" is deprecated, use "getRouter()" from the app function instead: "({ app, getRouter }) => { ... }"
   */
  public get router() {
    this.log.warn(
      new Deprecation(
        `[probot] "app.router" is deprecated, use "getRouter()" from the app function instead: "({ app, getRouter }) => { ... }"`
      )
    );

    return this.internalRouter;
  }

  /**
   * Get an {@link http://expressjs.com|express} router that can be used to
   * expose HTTP endpoints
   *
   * ```
   * module.exports = ({ app, getRouter }) => {
   *   // Get an express router to expose new HTTP endpoints
   *   const router = getRouter('/my-app');
   *
   *   // Use any middleware
   *   router.use(require('express').static(__dirname + '/public'));
   *
   *   // Add a new route
   *   router.get('/hello-world', (req, res) => {
   *     res.end('Hello World');
   *   });
   * };
   * ```
   *
   * @param path - the prefix for the routes* @param path
   *
   * @deprecated "app.route()" is deprecated, use the "getRouter()" argument from the app function instead: "({ app, getRouter }) => { ... }"
   */
  route(path?: string) {
    this.log.warn(
      new Deprecation(
        `[probot] "app.route()" is deprecated, use the "getRouter()" argument from the app function instead: "({ app, getRouter }) => { ... }"`
      )
    );

    return getRouter(this.internalRouter, path);
  }
}
