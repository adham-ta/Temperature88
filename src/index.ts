import { Logger } from "pino";

import { Application } from "./application";
import { Context, WebhookPayloadWithRepository } from "./context";
import { Options } from "./types";
import { Probot } from "./probot";
import { Server } from "./server/server";
import { ProbotOctokit } from "./octokit/probot-octokit";
import { run } from "./run";
import { getOptions } from "./get-options";
import { createNodeMiddleware } from "./create-node-middleware";
import { createProbot } from "./create-probot";

export {
  Logger,
  Context,
  Application,
  ProbotOctokit,
  run,
  Probot,
  Server,
  getOptions,
  createNodeMiddleware,
  createProbot,
};

/** NOTE: exported types might change at any point in time */
export { Options, WebhookPayloadWithRepository };
