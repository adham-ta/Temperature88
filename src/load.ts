import { Deprecation } from "deprecation";
import { Router } from "express";

import { Probot } from "./index";
import { Application } from "./application";
import { ApplicationFunction, ApplicationFunctionOptions } from "./types";
import { getRouter } from "./get-router";
import { resolveAppFunction } from "./helpers/resolve-app-function";

type DeprecatedKey =
  | "auth"
  | "load"
  | "log"
  | "on"
  | "receive"
  | "route"
  | "router";

const DEPRECATED_APP_KEYS: DeprecatedKey[] = [
  "auth",
  "load",
  "log",
  "on",
  "receive",
  "route",
  "router",
];
let didDeprecate = false;

function bindMethod(app: Probot | Application, key: keyof Application) {
  return typeof app[key] === "function" ? app[key].bind(app) : app[key];
}

/**
 * Loads an ApplicationFunction into the current Application
 * @param appFn - Probot application function to load
 */
export async function load(
  app: Application | Probot,
  router: Router | null,
  appFn: string | ApplicationFunction | ApplicationFunction[]
) {
  const deprecatedApp = DEPRECATED_APP_KEYS.reduce(
    (api: Record<string, unknown>, key: DeprecatedKey) => {
      Object.defineProperty(api, key, {
        get() {
          if (didDeprecate) return bindMethod(app, key);

          app.log.warn(
            new Deprecation(
              '[probot] "(app) => {}" is deprecated. Use "({ app }) => {}" instead'
            )
          );
          didDeprecate = true;

          return bindMethod(app, key);
        },
      });

      return api;
    },
    {}
  );

  if (Array.isArray(appFn)) {
    for (const fn of appFn) {
      await load(app, router, fn);
    }
    return app;
  }

  const fn = typeof appFn === "string" ? resolveAppFunction(appFn) : appFn;

  await fn(
    (Object.assign(deprecatedApp, {
      app,
      getRouter: getRouter.bind(null, router || app.router),
    }) as unknown) as ApplicationFunctionOptions
  );

  return app;
}
