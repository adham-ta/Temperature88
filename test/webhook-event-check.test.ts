import nock from "nock";
import { Probot } from "../src";

const appId = 1;
const privateKey = `-----BEGIN RSA PRIVATE KEY-----
MIIBOQIBAAJBAIILhiN9IFpaE0pUXsesuuoaj6eeDiAqCiE49WB1tMB8ZMhC37kY
Fl52NUYbUxb7JEf6pH5H9vqw1Wp69u78XeUCAwEAAQJAb88urnaXiXdmnIK71tuo
/TyHBKt9I6Rhfzz0o9Gv7coL7a537FVDvV5UCARXHJMF41tKwj+zlt9EEUw7a1HY
wQIhAL4F/VHWSPHeTgXYf4EaX2OlpSOk/n7lsFtL/6bWRzRVAiEArzJs2vopJitv
A1yBjz3q2nX+zthk+GLXrJQkYOnIk1ECIHfeFV8TWm5gej1LxZquBTA5pINoqDVq
NKZSuZEHqGEFAiB6EDrxkovq8SYGhIQsJeqkTMO8n94xhMRZlFmIQDokEQIgAq5U
r1UQNnUExRh7ZT0kFbMfO9jKYZVlQdCL9Dn93vo=
-----END RSA PRIVATE KEY-----`;

// tslint:disable-next-line
const noop = async () => {};

/**
 * Returns a mocked request for `/meta` with the subscribed `events`.
 *
 * By default, the mocked payload indicates the a GitHub App is subscribed to
 * the `issues` event.
 */
function mockAppMetaRequest(events: string[] = ["issues"]) {
  return { events };
}

jest.setTimeout(10000);

describe("webhook-event-check", () => {
  let originalNodeEnv: string;
  let originalJestWorkerId: string | undefined;

  beforeAll(() => {
    originalNodeEnv = process.env.NODE_ENV || "test";
    originalJestWorkerId = process.env.JEST_WORKER_ID;
    nock.disableNetConnect();
  });

  afterAll(() => {
    nock.restore();
  });

  beforeEach(() => {
    // We need to re-configure environment variables to avoid
    // webhook-event-check from triggering its smart-disable feature.
    delete process.env.DISABLE_WEBHOOK_EVENT_CHECK;
    delete process.env.JEST_WORKER_ID;
    process.env.NODE_ENV = "development";
    nock.cleanAll();
  });

  afterEach(() => {
    process.env.NODE_ENV = originalNodeEnv;
    process.env.JEST_WORKER_ID = originalJestWorkerId;
  });

  test("caches result of /app, logs error for event that the app is not subscribed to", async () => {
    nock("https://api.github.com")
      .get("/app")
      .reply(200, mockAppMetaRequest(["label", "star"]));

    const app = new Probot({ appId, privateKey });

    const spyOnLogError = jest.spyOn(app.log, "error");

    app.on("label.edited", noop);
    app.on("label.deleted", noop);
    app.on("team.created", noop);

    // let's give the event check a moment to send its request
    await new Promise((resolve) => setTimeout(resolve, 2500));

    expect(spyOnLogError).toHaveBeenCalledTimes(1);
    expect(spyOnLogError).toMatchSnapshot();
  });

  test("logs no error when listening with app.webhooks.onAny", async () => {
    nock("https://api.github.com")
      .get("/app")
      .reply(200, mockAppMetaRequest([]));

    const probot = new Probot({ appId, privateKey });
    let spyOnLogError;

    probot.load(async ({ app }) => {
      spyOnLogError = jest.spyOn(app.log, "error");

      app.webhooks.onAny(noop);
    });

    // let's give the event check a moment to send its request
    await new Promise((resolve) => setTimeout(resolve, 500));

    expect(spyOnLogError).not.toHaveBeenCalled();
  });

  describe("warn user when", () => {
    test("listening to unsubscribed event", async () => {
      nock("https://api.github.com")
        .defaultReplyHeaders({ "Content-Type": "application/json" })
        .get("/app")
        .reply(200, mockAppMetaRequest());

      const app = new Probot({ appId, privateKey });
      const spyOnLogError = jest.spyOn(app.log, "error");

      app.on("pull_request.opened", noop);

      // let's give the event check a moment to send its request
      await new Promise((resolve) => setTimeout(resolve, 500));

      expect(spyOnLogError).toHaveBeenCalledTimes(1);
      expect(spyOnLogError).toMatchSnapshot();
    });

    test("unable to retrieve app metadata", async () => {
      nock("https://api.github.com")
        .defaultReplyHeaders({ "Content-Type": "application/json" })
        .get("/app")
        .reply(404);

      const app = new Probot({ appId, privateKey });
      let spyOnLogError;

      spyOnLogError = jest.spyOn(app.log, "error");

      app.on("pull_request.opened", noop);

      // let's give the event check a moment to send its request
      await new Promise((resolve) => setTimeout(resolve, 500));

      expect(spyOnLogError).toHaveBeenCalledTimes(1);
      expect(spyOnLogError).toMatchSnapshot();
    });
  });

  describe("can be disabled", () => {
    beforeEach(() => {
      delete process.env.DISABLE_WEBHOOK_EVENT_CHECK;
      delete process.env.JEST_WORKER_ID;
      delete process.env.NODE_ENV;
    });

    test("when JEST_WORKER_ID is set", async () => {
      process.env.JEST_WORKER_ID = "mocked_id";

      const app = new Probot({ appId, privateKey });

      const spyOnLogError = jest.spyOn(app.log, "error");
      app.on("pull_request.opened", noop);

      // let's give app setup a moment
      await new Promise((resolve) => setTimeout(resolve, 500));

      expect(spyOnLogError).not.toHaveBeenCalled();
    });

    test("when DISABLE_WEBHOOK_EVENT_CHECK is true", async () => {
      process.env.DISABLE_WEBHOOK_EVENT_CHECK = "true";

      const app = new Probot({ appId, privateKey });
      const spyOnLogError = jest.spyOn(app.log, "error");
      app.on("pull_request.opened", noop);

      // let's give app setup a moment
      await new Promise((resolve) => setTimeout(resolve, 500));

      expect(spyOnLogError).not.toHaveBeenCalled();
    });

    test("when NODE_ENV is production", async () => {
      process.env.NODE_ENV = "production";

      const app = new Probot({ appId, privateKey });
      const spyOnLogError = jest.spyOn(app.log, "error");
      app.on("pull_request.opened", noop);

      // let's give app setup a moment
      await new Promise((resolve) => setTimeout(resolve, 500));

      expect(spyOnLogError).not.toHaveBeenCalled();
    });

    test('when NODE_ENV starts with "test"', async () => {
      process.env.NODE_ENV = "testing";

      const app = new Probot({ appId, privateKey });
      const spyOnLogError = jest.spyOn(app.log, "error");
      app.on("pull_request.opened", noop);

      // let's give app setup a moment
      await new Promise((resolve) => setTimeout(resolve, 500));

      expect(spyOnLogError).not.toHaveBeenCalled();
    });
  });
});
