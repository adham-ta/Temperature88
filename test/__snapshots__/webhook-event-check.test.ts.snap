// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`webhook-event-check caches result of /app, logs error for event that the app is not subscribed to 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "Your app is attempting to listen to \\"team.created\\", but your GitHub App is not subscribed to the \\"team\\" event.",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`webhook-event-check warn user when listening to unsubscribed event 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "Your app is attempting to listen to \\"pull_request.opened\\", but your GitHub App is not subscribed to the \\"pull request\\" event.",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`webhook-event-check warn user when unable to retrieve app metadata 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "Your app is attempting to listen to \\"pull_request.opened\\", but your GitHub App is not subscribed to the \\"pull request\\" event.",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
